test/hooks.test.js

```diff
@@ -664,6 +664,49 @@ test('onRoute hook should preserve handler function in options of shorthand rout
   })
 })
 
+// issue ref https://github.com/fastify/fastify-compress/issues/140
+test('onRoute hook should be called once when prefixTrailingSlash', t => {
+  t.plan(3)
+
+  let onRouteCalled = 0
+  let routePatched = 0
+
+  const fastify = Fastify({ ignoreTrailingSlash: false })
+
+  // a plugin that patches route options, similar to fastify-compress
+  fastify.register(fp(function myPlugin (instance, opts, next) {
+    function patchTheRoute () {
+      routePatched++
+    }
+
+    instance.addHook('onRoute', function (routeOptions) {
+      onRouteCalled++
+      patchTheRoute(routeOptions)
+    })
+
+    next()
+  }))
+
+  fastify.register(function routes (instance, opts, next) {
+    instance.route({
+      method: 'GET',
+      url: '/',
+      prefixTrailingSlash: 'both',
```

Comment:
Not sure I understand @AyoubElk. Currently I get `onRoute` called for these routes.

```js
[ 'GET /v1', 'HEAD /v1/v1', 'HEAD /v1' ]
```

And If I revert [this](https://github.com/fastify/fastify/pull/2782/files#diff-cf6996cafd1df0bc168a5c5f3a7470f11fc3aea9f908091d5dd6be2e7b30a2c3R199) and make calling of `onRoute` unconditional again, onRoute gets called for following routes.

```js
[ 'GET /v1', 'HEAD /v1/v1', 'GET /v1/', 'HEAD /v1', 'HEAD /v1/' ]
```

Which seems correct to me as `onRoute` is only called for route (GET as well as HEAD) without trailing slash as intended.

Now about that mysterious `HEAD /v1/v1` route. I checked in `master` and it seems to be a pre-existing bug.
@mcollina

Reply:
> Not sure I understand @AyoubElk. Currently I get `onRoute` called for these routes.
> 
> ```js
> [ 'GET /v1', 'HEAD /v1/v1', 'HEAD /v1' ]
> ```
> 

@salmanm You have to add the `exposeHeadRoute` option to the route definition.

The reason is that `prepareRoute` is called for the `HEAD` with the already mutated route definition, and the `onRoute` hook will be called again and patch it again.

This is the updated test that displays this behavior:

```js
// issue ref https://github.com/fastify/fastify-compress/issues/140
test('onRoute hook should be called once when prefixTrailingSlash', t => {
  t.plan(3)

  let onRouteCalled = 0
  let routePatched = 0

  const fastify = Fastify({ ignoreTrailingSlash: false })

  // a plugin that patches route options, similar to fastify-compress
  fastify.register(fp(function myPlugin (instance, opts, next) {
    function patchTheRoute () {
      routePatched++
    }

    instance.addHook('onRoute', function (routeOptions) {
      onRouteCalled++
      patchTheRoute(routeOptions)
    })

    next()
  }))

  fastify.register(function routes (instance, opts, next) {
    instance.route({
      method: 'GET',
      url: '/',
      exposeHeadRoute: true,
      prefixTrailingSlash: 'both',
      handler: (req, reply) => {
        reply.send({ hello: 'world' })
      }
    })

    next()
  }, { prefix: '/prefix' })

  fastify.ready(err => {
    t.error(err)
    t.is(onRouteCalled, 1) // onRoute hook was called once
    t.is(routePatched, 1) // and plugin acted once and avoided redundaunt route patching
  })
})
```

Output:
```
❯ tap -J test/hooks.test.js 
 FAIL  test/hooks.test.js
 ✖ should be equal

  test/hooks.test.js                                                                              
  704 |   fastify.ready(err => {                                                                  
  705 |     t.error(err)                                                                          
> 706 |     t.is(onRouteCalled, 1) // onRoute hook was called once                                
      | ------^                                                                                   
  707 |     t.is(routePatched, 1) // and plugin acted once and avoided redundaunt route patching  
  708 |   })                                                                                      
  709 | })                                                                                        

  --- expected    
  +++ actual      
  @@ -1,1 +1,1 @@ 
  -1              
  +3              

  test: test/hooks.test.js onRoute hook should be called once when prefixTrailingSlash
  stack: |
    cb (test/hooks.test.js:706:7)
    cb (fastify.js:438:11)
    exit (lib/hooks.js:90:5)
    manageTimeout (lib/hooks.js:107:11)
    _encapsulateThreeParam (node_modules/avvio/boot.js:545:7)
    Boot.timeoutCall (node_modules/avvio/boot.js:447:5)
    Boot.callWithCbOrNextTick (node_modules/avvio/boot.js:428:19)
    Task.release (node_modules/fastq/queue.js:143:16)
    worked (node_modules/fastq/queue.js:195:10)
    Boot.done (node_modules/avvio/boot.js:456:7)
    fastify.js:429:9
    Object._encapsulateThreeParam (node_modules/avvio/boot.js:545:7)
    Boot.timeoutCall (node_modules/avvio/boot.js:447:5)
    Boot.callWithCbOrNextTick (node_modules/avvio/boot.js:428:19)
    release (node_modules/fastq/queue.js:143:16)
    Object.resume (node_modules/fastq/queue.js:76:7)
    node_modules/avvio/boot.js:167:18
    node_modules/avvio/plugin.js:269:7
    done (node_modules/avvio/plugin.js:201:5)
    check (node_modules/avvio/plugin.js:225:9)

 FAIL  test/hooks.test.js
 ✖ should be equal

  test/hooks.test.js                                                                              
  705 |     t.error(err)                                                                          
  706 |     t.is(onRouteCalled, 1) // onRoute hook was called once                                
> 707 |     t.is(routePatched, 1) // and plugin acted once and avoided redundaunt route patching  
      | ------^                                                                                   
  708 |   })                                                                                      
  709 | })                                                                                        
  710 |                                                                                           

  --- expected    
  +++ actual      
  @@ -1,1 +1,1 @@ 
  -1              
  +3              

  test: test/hooks.test.js onRoute hook should be called once when prefixTrailingSlash
  stack: |
    cb (test/hooks.test.js:707:7)
    cb (fastify.js:438:11)
    exit (lib/hooks.js:90:5)
    manageTimeout (lib/hooks.js:107:11)
    _encapsulateThreeParam (node_modules/avvio/boot.js:545:7)
    Boot.timeoutCall (node_modules/avvio/boot.js:447:5)
    Boot.callWithCbOrNextTick (node_modules/avvio/boot.js:428:19)
    Task.release (node_modules/fastq/queue.js:143:16)
    worked (node_modules/fastq/queue.js:195:10)
    Boot.done (node_modules/avvio/boot.js:456:7)
    fastify.js:429:9
    Object._encapsulateThreeParam (node_modules/avvio/boot.js:545:7)
    Boot.timeoutCall (node_modules/avvio/boot.js:447:5)
    Boot.callWithCbOrNextTick (node_modules/avvio/boot.js:428:19)
    release (node_modules/fastq/queue.js:143:16)
    Object.resume (node_modules/fastq/queue.js:76:7)
    node_modules/avvio/boot.js:167:18
    node_modules/avvio/plugin.js:269:7
    done (node_modules/avvio/plugin.js:201:5)
    check (node_modules/avvio/plugin.js:225:9)

 FAIL  test/hooks.test.js 2 failed of 546 1s
 ✖ should be equal
 ✖ should be equal


                         
  🌈 SUMMARY RESULTS 🌈  
                         

 FAIL  test/hooks.test.js 2 failed of 546 1s
 ✖ should be equal
 ✖ should be equal

Suites:   1 failed, 1 of 1 completed
Asserts:  2 failed, 544 passed, of 546
Time:     2s
-----------------------------|----------|----------|----------|----------|-------------------|
File                         |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
-----------------------------|----------|----------|----------|----------|-------------------|
All files                    |    68.43 |    57.92 |    63.89 |    69.55 |                   |
 fastify                     |    63.27 |    59.26 |    58.82 |    63.33 |                   |
  fastify.js                 |    63.27 |    59.26 |    58.82 |    63.33 |... 12,614,617,631 |
 fastify/lib                 |    69.25 |    57.73 |    64.98 |    70.56 |                   |
  contentTypeParser.js       |    69.66 |    56.52 |       70 |    72.79 |... 39,253,265,274 |
  context.js                 |    76.67 |    66.67 |       50 |    79.31 | 32,33,35,36,37,39 |
  decorate.js                |       60 |    40.91 |       60 |    61.76 |... 63,67,68,69,90 |
  errors.js                  |      100 |      100 |      100 |      100 |                   |
  fourOhFour.js              |    59.15 |    52.78 |    66.67 |    60.87 |... 65,167,168,169 |
  handleRequest.js           |    77.05 |     80.7 |      100 |    78.33 |... 94,126,127,134 |
  headRoute.js               |       20 |     12.5 |    33.33 |    14.29 |... 14,17,19,21,26 |
  hooks.js                   |    74.17 |    64.81 |    78.95 |    75.21 |... 07,220,225,229 |
  initialConfigValidation.js |       75 |       60 |    66.67 |    78.95 |       11,12,13,26 |
  logger.js                  |    77.19 |    61.54 |     87.5 |    76.36 |... 39,40,50,74,78 |
  pluginOverride.js          |    95.12 |       85 |      100 |       95 |             53,73 |
  pluginUtils.js             |    74.63 |    47.06 |    78.57 |    77.36 |... 79,81,82,83,97 |
  reply.js                   |    71.03 |    56.54 |    71.11 |    71.25 |... 00,701,702,704 |
  reqIdGenFactory.js         |      100 |      100 |      100 |      100 |                   |
  request.js                 |    38.81 |     3.57 |    23.08 |    40.63 |... 44,149,154,159 |
  route.js                   |       82 |    77.22 |    84.21 |    82.65 |... 26,447,452,456 |
  schema-compilers.js        |    81.48 |    66.67 |    71.43 |    84.62 |       17,41,42,49 |
  schemas.js                 |    61.29 |    46.15 |    69.23 |    63.33 |... 86,102,109,118 |
  server.js                  |     56.7 |    41.94 |     37.5 |    57.89 |... 57,162,163,168 |
  symbols.js                 |      100 |      100 |      100 |      100 |                   |
  validation.js              |    48.33 |       44 |    55.56 |    52.73 |... 05,106,107,108 |
  warnings.js                |      100 |      100 |      100 |      100 |                   |
  wrapThenable.js            |       40 |    46.15 |    66.67 |       40 |... 34,35,38,39,40 |
-----------------------------|----------|----------|----------|----------|-------------------|

```

Reply:
> Now about that mysterious `HEAD /v1/v1` route. I checked in `master` and it seems to be a pre-existing bug.
> @mcollina

The `HEAD /v1/v1` route is caused by the same issue, mutating the route definition for `GET` route, then prefixing it again when `afterRouteAdded` is called for the `HEAD` route

Reply:
I have tried with the same config, both of them are in route options.

```js
exposeHeadRoute: true,
prefixTrailingSlash: 'both',
```

In your failing tests, the onRoute is being called thrice, it's the same as I mentioned in my previous comment. Once for GET and then for HEAD (not talking about that /v1/v1 route for the moment, which is a separate bug)
The patching of routes aren't always unconditional as shown in the simple test I added. The plugin would check for the relevant method before patching routes.

So again, I don't quite understand what the problem is here.

Reply:
Yes you're right, I made a mistake updating the test. 
The issue appears when mutating the route definition, which is something you don't do in your test.

Will provide a reproduction example shortly.

Reply:
FYI allowing to mutate route options isn't in the intent of this PR.

Reply:
> FYI allowing to mutate route options isn't in the intent of this PR.

Yes I know, however in fastify currently the route definition can be mutated by an `onRoute` hook then passed to the next ones.

You can check this test to understand what I mean:

```js
test('mutating the route definition in onRoute hooks shouldnt interfere with afterRouteAdded() being called twice', t => {
  t.plan(3)

  const routeObject = {
    handler: (req, reply) => { reply.send('hi from handler') },
    method: 'GET',
    exposeHeadRoute: true,
    prefixTrailingSlash: 'both',
    url: '/'
  }

  const fastify = Fastify()
  t.tearDown(() => fastify.close())

  let onRequestCallCounter = 0

  fastify.register(async function (f) {
    f.addHook('onRoute', (routeOptions) => {
      if (routeOptions.onRequest) {
        routeOptions.onRequest = [onRequest].concat(routeOptions.onRequest)
      } else {
        routeOptions.onRequest = [onRequest]
      }
      function onRequest (req, reply, next) {
        onRequestCallCounter++
        next()
      }
    })

    f.route(routeObject)
  }, { prefix: '/v1' })

  fastify.listen(0, err => {
    t.error(err)
    sget({
      method: 'HEAD',
      url: 'http://localhost:' + fastify.server.address().port + '/v1/'
    }, (err, response, body) => {
      t.error(err)
      t.equal(onRequestCallCounter, 1)
    })
  })
})
```

Output:

```
 FAIL  test/route.test.js
 ✖ should be equal

  test/route.test.js                                         
  871 |       t.error(err)                                   
  872 |       // The onRequest hook was added & called once  
> 873 |       t.equal(onRequestCallCounter, 1)               
      | --------^                                            
  874 |     })                                               
  875 |   })                                                 
  876 | })                                                   

  --- expected    
  +++ actual      
  @@ -1,1 +1,1 @@ 
  -1              
  +2              
```

Reply:
There seems to be a gap in expectation here. I am not sure how this is relevant.

Your above test gives `onRequestCallCounter === 2` on this branch and `onRequestCallCounter === 4` on `master`. 

Are you sure there's a bug? Coz I fail to see one.

Reply:
It should be called only once.

Reply:
Well, routeOptions are shallow cloned in route.js and you're mutating a nested key so onRoute is set twice for GET. 

I think may be it's a bug in test 😉

Nonetheless, it's not related to the changes in this PR, so if you're really sure there's a bug that needs fixing, please create a new issue.

Reply:
I assumed it's related because the main issue this PR solves is that fastify in some cases creates more than one route from the same user provided route definition object (prefixed routes/`HEAD` routes when `exposeHeadRoute` is used).
And since it's possible for `onRoute` hooks to mutate that definition, sometimes lifecycle hooks are added more than once to the same object and cause issues (as displayed here in [fastify-compress](https://github.com/fastify/fastify-compress/pull/141) where data is corrupted because `onSend` was added/called twice).

This PR solves the issue for prefixed routes, but for `HEAD` routes it still exists. Your test doesn't display that because it increments the `routePatched` counter but doesn't mutate the route definition (similar to what `fastify-compress` does).

Anyway, I will look more into it and move the discussion to a separate issue when I have some spare time.

Thanks.